open_spiel/games/solitaire.cc

```diff
@@ -0,0 +1,1301 @@
+        switch (action_id) {
+            case kDraw : {
+                return "kDraw";
+            }
+            case kRevealAs ... kRevealKd : {
+                auto revealed_card = Card((int) action_id);
+                std::string result;
+                absl::StrAppend(&result, "kReveal", revealed_card.ToString(is_colored));
+                return result;
+            }
+            case kMove__Ks ... kMoveKdQc : {
+                auto move = Move(action_id);
+                return move.ToString(is_colored);
+            }
+            default : {
+                return "Missing Action";
+            }
+        }
+    }
+
+    std::string             SolitaireState::InformationStateString(Player player) const {
+        SPIEL_CHECK_GE(player, 0);
+        SPIEL_CHECK_LT(player, num_players_);
+        return HistoryString();
+    }
+
+    std::string             SolitaireState::ObservationString(Player player) const {
+        SPIEL_CHECK_GE(player, 0);
+        SPIEL_CHECK_LT(player, num_players_);
+        return ToString();
+    }
+
+    void                    SolitaireState::ObservationTensor(Player player, std::vector<double> *values) const {
+        SPIEL_CHECK_GE(player, 0);
+        SPIEL_CHECK_LT(player, num_players_);
+
+        values->resize(game_->ObservationTensorSize());
+        std::fill(values->begin(), values->end(), 0.0);
+        auto ptr = values->begin();
+
+        for (auto & foundation : foundations) {
+            if (foundation.cards.empty()) {
+                ptr[0] = 1;
+            } else {
+                auto last_rank = foundation.cards.back().rank;
+                if (last_rank >= kA && last_rank <= kK) {
+                    ptr[last_rank] = 1;
+                }
+            }
+            ptr += FOUNDATION_TENSOR_LENGTH;
+        }
+
+        for (auto & tableau : tableaus) {
+            if (tableau.cards.empty()) {
+                ptr[7] = 1.0;
+                continue;
+            } else {
+                int num_hidden_cards = 0;
+                for (auto & card : tableau.cards) {
+                    if (card.hidden && num_hidden_cards <= MAX_HIDDEN_CARDS) {
+                        ptr[num_hidden_cards] = 1.0;
+                        ++num_hidden_cards;
+                    } else {
+                        auto tensor_index = card.GetIndex() + MAX_HIDDEN_CARDS;
+                        ptr[tensor_index] = 1.0;
+                    }
+                }
+            }
+            ptr += TABLEAU_TENSOR_LENGTH;
+        }
+
+        if (waste.cards.empty()) {
+            return;
+        } else {
+            for (auto & card : waste.cards) {
+                if (card.hidden) {
+                    ptr[0] = 1;
+                } else {
+                    auto tensor_index = card.GetIndex() + 1;
+                    ptr[tensor_index] = 1.0;
+                }
+                ptr += WASTE_TENSOR_LENGTH;
+            }
+        }
+
+        /*
```

Comment:
Please don't leave commented-out code in place.
If you really want to keep it, it would be better to have a flag for it, so that the compiler sees it.